File: base/science-tech-maths/programming/algorithms/job-interviews/snippets/reverse_singly_linked_list.cpp
#include <iostream>
#include <memory>

template <typename T>
class Node {
   public:
    T m_data;
    Node* p_next_node;

    Node(T data) : m_data(data), p_next_node(nullptr) {}

    // operator << is not a member function, but needs access to member parameters
    // we define it as friend, even though it's useless here because everything is public...
    friend std::ostream& operator<<(std::ostream& out, const Node& node) {
        if (node.p_next_node) {
            return out << node.m_data << " -> " << *node.p_next_node;
        }
        return out << node.m_data;
    }
};

int main() {
    Node node_1(1);
    Node node_2(2);
    Node node_3(3);
    Node node_4(4);

    node_1.p_next_node = &node_2;
    node_2.p_next_node = &node_3;
    node_3.p_next_node = &node_4;

    std::cout << node_1 << std::endl;

    // reverse list
    decltype(node_1)* prev = nullptr;  //    Inspects the declared type of an entity.
    auto* head = &node_1;
    // int z = std::exchange(x, y);
    // x is assigned the value of y,
    // z is assigned the value that x had initially.
    // 1st call: prev is assigned head, returns prev
    // 2nd call: head->p_next_node is assigned prev, the current node has been linked to the previous one
    // head is set to head->p_next_node (next node in the list to revert)
    while (head != nullptr) {
        head = std::exchange(head->p_next_node, std::exchange(prev, head));
    }

    std::cout << node_4 << std::endl;

    return 0;
}

File: base/science-tech-maths/programming/algorithms/job-interviews/snippets/reverse_singly_linked_list.py
class Node:
    def __init__(self, data, next_node: "Node" | "None" = None):
        self.data = data
        self.next_node = next_node

    def __str__(self):
        if self.next_node:
            return " -> ".join((str(self.data), str(self.next_node)))
        return str(self.data)


def main():
    linked_list = Node(1, Node(2, Node(3)))
    print(linked_list)

    prev_ = None
    curr_ = linked_list
    next_ = curr_.next_node

    while next_:
        next_ = curr_.next_node
        curr_.next_node = prev_
        prev_ = curr_
        curr_ = next_

    # second version:
    # while curr_:
    #     curr_.next_node = prev_
    #     prev_ = curr_
    #     curr_ = next_
    #     if not curr_:
    #         break
    #     next_ = curr_.next_node

    linked_list = prev_
    print(linked_list)


if __name__ == "__main__":
    main()
